### PR TITLE
OSDOCS-2688: Conceptual information about HPA

### PR DESCRIPTION
**Pull request by Amrita42**
- Applies to 4.10+
- [Jira](https://issues.redhat.com/browse/OSDOCS-2688)
- [Preview](https://deploy-preview-40704--osdocs.netlify.app/openshift-enterprise/latest/nodes/pods/nodes-pods-autoscaling.html#nodes-pods-autoscaling-more-about_nodes-pods-autoscaling)

Queries I have:
1) We currently have six parameters in K8S docs:
--horizontal-pod-autoscaler-sync-period --horizontal-pod-autoscaler-downscale-stabilization    --horizontal-pod-autoscaler-cpu-initialization-period duration,  --horizontal-pod-autoscaler-initial-readiness-delay duration   --horizontal-pod-autoscaler-sync-period duration .
All these parameters I believe are in  `stablizationWindowSeconds`.  Can this be confirmed? 

2) Best practices for the HPA , need more insights on this. 
---------------------------------------------------------------------------------------------------------------
Reviewers
Please share your feedback/perspective. Also, do we need to add more conceptual information for example do we need an in-depth analysis of the Algorithm workflow? Would this be too much information for the user?

@AlexonOliveiraRH   @gauravsingh85  @joelsmith @wangchen615  @sjstout @weinliu

### CODE REVIEW

**Comment by AlexonOliveiraRH** [comment]
> * Applies to 4.10+
> * [Jira](https://issues.redhat.com/browse/OSDOCS-2688)
> * [Preview](https://deploy-preview-40704--osdocs.netlify.app/openshift-enterprise/latest/nodes/pods/nodes-pods-autoscaling.html)
> 
> Queries I have:
> 
> 1. We currently have six parameters in K8S docs:
>    --horizontal-pod-autoscaler-sync-period --horizontal-pod-autoscaler-downscale-stabilization    --horizontal-pod-autoscaler-cpu-initialization-period duration,  --horizontal-pod-autoscaler-initial-readiness-delay duration   --horizontal-pod-autoscaler-sync-period duration .
>    All these parameters I believe are in  `stablizationWindowSeconds`.  Can this be confirmed?

@Amrita42, I guess they're meant for different things:

- **--horizontal-pod-autoscaler-sync-period:** Kubernetes implements horizontal pod autoscaling as a control loop that runs intermittently (it is not a continuous process). The interval is set by this parameter to the kube-controller-manager (and the default interval is 15 seconds).
- **--horizontal-pod-autoscaler-downscale-stabilization:** Right before HPA scales the target, the scale recommendation is recorded. The controller considers all recommendations within a configurable window choosing the highest recommendation from within that window. This value can be configured using this flag, which defaults to 5 minutes.
- **--horizontal-pod-autoscaler-cpu-initialization-period duration:** Once a pod has become ready, it considers any transition to ready to be the first if it occurred within a longer, configurable time since it started. This value is configured with this flag, and its default is 5 minutes.
- **--horizontal-pod-autoscaler-initial-readiness-delay duration:** Due to technical constraints, the HorizontalPodAutoscaler controller cannot exactly determine the first time a pod becomes ready when determining whether to set aside certain CPU metrics. Instead, it considers a Pod "not yet ready" if it's unready and transitioned to unready within a short, configurable window of time since it started. This value is configured with this flag, and its default is 30 seconds.

The `stabilizationWindowSeconds` is used to restrict the flapping of replicas count when the metrics used for scaling keep fluctuating. The autoscaling algorithm uses this window to infer a previous desired state and avoid unwanted changes to workload scale.

> 2. Best practices for the HPA , need more insights on this.
> 

I think this topic is very wide and I'm not sure we would be able to include all the best practices out there related to K8s HPA. I guess the two recommendations already provided are valid.

> Reviewers Please share your feedback/perspective. Also, do we need to add more conceptual information for example do we need an in-depth analysis of the Algorithm workflow? Would this be too much information for the user?
> 
> @AlexonOliveiraRH @rphillips @gauravsingh85 @joelsmith @wangchen615 @sjstout

In regards the content, kudos, because it seems richer and more complete than its previous versions. There's a _More about..._ topic with the main information well summarized in relation to the feature, and it's based on the official K8s documentation. Also, I don't think the algorithm part should be included in our documentation given the fact that the rest of the documentation is very clean, but at the same time it summarizes the important points and provides the necessary information about the functionality of the feature, mainly giving a more conceptual understanding of it that previous versions did not provide. I'll reread this preview a few more times more carefully to make sure I'm not missing anything, but so far it looks pretty good to me.

**Comment by Amrita42** [comment]
@AlexonOliveiraRH thanks for the feedback.

**Comment by Amrita42** [comment]
Added a section on requests and limits and a link to the blog on HPA from our Red Hat portal.

**Comment by weinliu** [comment]
I do not think we need to add more conceptual information on this, the doc now looks clean and easy to understand.

File: modules/nodes-pods-autoscaling-more-about.adoc [hunk 1]
```diff
@@ -0,0 +1,68 @@
+// Module included in the following assemblies:
+//
+// * nodes/nodes-pods-autoscaling-about.adoc
+
+:_content-type: CONCEPT
+[id="nodes-pods-autoscaling-more-about_{context}"]
+= More about horizontal pod autoscalers
+
+The horizontal pod autoscaler (HPA) extends the concept of pod auto-scaling. The HPA lets you create and manage a group of load-balanced nodes. The HPA automatically increases or decreases the number of pods when a given CPU or memory threshold is crossed.
```
Comment (adellape) [inline modules/nodes-pods-autoscaling-more-about.adoc:9]: Could this paragraph just be at the start of the "How does the HPA work" section? Versus having a separate "More about..." heading, considering there's already an "Understanding..." heading right before it.

E.g., with the TOC ultimately looking something like:

* Understanding horizontal pod autoscalers
  *	How does the HPA work?
  *	Supported metrics
  *	About requests and limits
  *	How to use resource metrics?
  *	Best practices
  *	Scaling practices

You'd probably want to break this up into separate modules to achieve that, though.

File: modules/nodes-pods-autoscaling-more-about.adoc [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+== How does the HPA work?
```
Comment (adellape) [inline modules/nodes-pods-autoscaling-more-about.adoc:11]: Each section heading should have a unique ID/anchor added. Here and throughout for the other sections.

File: modules/nodes-pods-autoscaling-more-about.adoc [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+
+.High level workflow of the HPA
+image::HPAflow.png[workflow]
+
+The HPA is an API resource in the Kubernetes autoscaling API group. The autoscaler works as a control loop with a default of 15 seconds for the sync period. During this period, the controller manager queries the CPU and/or the memory utilization against what is defined in the YAML file for the HPA.
```
Comment (adellape) [inline modules/nodes-pods-autoscaling-more-about.adoc:16]: See IBMSG for guidance re: "and/or"

File: modules/nodes-pods-autoscaling-more-about.adoc [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+
+== Best practices
+
+.All pods must have resource requests configured
+The HPA makes a scaling decision based on the observed CPU or memory utilization values of pods in an {product-title} cluster. Utilization values are calculated as a percentage of the resource requests of each pod.
+Missing resource request values can affect the optimal performance of the HPA.
+
+.Configure the cool down period
+During horizontal pod autoscaling, there might be a rapid scaling of events without a time gap. Configure the cool down period to prevent frequent replica fluctuations.
+You can specify a cool down period by confiuring `stabilizationWindowSeconds`. The stabilization window is used to restrict the the fluctuation of replicas count when the metrics used for scaling keep fluctuating.
```
Comment (adellape) [inline modules/nodes-pods-autoscaling-more-about.adoc:56]:
```
by configuring the `stabilizationWindowSeconds` field.
```

also note "confiuring" typo

File: modules/nodes-pods-autoscaling-more-about.adoc [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+The autoscaling algorithm uses this window to infer a previous desired state and avoid unwanted changes to workload scale.
+
+For example, a stabilization window is specified for scaleDown:
```
Comment (adellape) [inline modules/nodes-pods-autoscaling-more-about.adoc:59]:
```
for the `scaleDown` behavior:
```

or

```
for the `scaleDown` field:
```

?

File: nodes/pods/nodes-pods-autoscaling.adoc [hunk 6]
```diff
@@ -41,5 +43,7 @@ include::modules/nodes-pods-autoscaling-status-viewing.adoc[leveloffset=+2]
 
 == Additional resources
 
-For more information on replication controllers and deployment controllers,
+* For more information on replication controllers and deployment controllers,
 see xref:../../applications/deployments/what-deployments-are.adoc#what-deployments-are[Understanding deployments and deployment configs].
+
+* For an example on the usage of HPA, see https://cloud.redhat.com/blog/horizontal-pod-autoscaling-of-quarkus-application-based-on-memory-utilization.
```
Comment (adellape) [inline nodes/pods/nodes-pods-autoscaling.adoc:49]: If you wanted to hide the URL and use the blog title:

```
see link:https://cloud.redhat.com/blog/horizontal-pod-autoscaling-of-quarkus-application-based-on-memory-utilization[Horizontal Pod Autoscaling of Quarkus Application Based on Memory Utilization].
```

File: modules/nodes-pods-autoscaling-more-about.adoc (hunk 2 above)
Comment (Amrita42) [reply modules/nodes-pods-autoscaling-more-about.adoc:11]: Yeah I missed that , good catch!

File: modules/nodes-pods-autoscaling-more-about.adoc (hunk 5 above)
Comment (Amrita42) [reply modules/nodes-pods-autoscaling-more-about.adoc:59]: second option is correct

File: modules/nodes-pods-autoscaling-more-about.adoc (hunk 1 above)
Comment (Amrita42) [reply modules/nodes-pods-autoscaling-more-about.adoc:9]: Going with your suggestion.
I have kept About requests and limits and How to use resource metrics? as one conceptual module as they are interelated.

Comment (Amrita42) [reply modules/nodes-pods-autoscaling-more-about.adoc:9]: Supported Metrics is part of the Understanding horizontal pod auto scaler module.  So the TOC would look like:

Understanding horizontal pod autoscalers (part of Supported metrics)
How does the HPA work?
About requests and limits (including how to use resource metrics?)
Best practices
Scaling practices

**Pull request merged**